File: pusheen/closed_loop/frontend/src/lib/summarizer.ts
/**
 * AI summarization & anti-clickbait title generation.
 * Uses Gemini Flash via its OpenAI-compatible endpoint.
 */
import OpenAI from 'openai';
import { cacheGet, cacheSet, makeCacheKey } from './cache';
import { Article, extractArticleContent } from './news-fetcher';

let geminiClient: OpenAI | null = null;

function getGeminiClient(): OpenAI {
  if (!geminiClient) {
    geminiClient = new OpenAI({
      apiKey: process.env.GEMINI_API_KEY || '',
      baseURL:
        process.env.GEMINI_BASE_URL ||
        'https://generativelanguage.googleapis.com/v1beta/openai/',
    });
  }
  return geminiClient;
}

const SYSTEM_PROMPT = `You are a news analyst. Given an article's original title and content, produce:

1. **generated_title** — A concise, accurate, non-clickbait title that faithfully represents the article's actual content. Remove sensationalism, vague teasers, and emotional manipulation. Keep it informative.
2. **summary** — A TL;DR of 2-3 sentences capturing the key facts. No fluff.
3. **clickbait_score** — A float from 0.0 (not clickbait) to 1.0 (extreme clickbait) rating how clickbaity the ORIGINAL title is.
4. **is_clickbait** — Boolean, true if clickbait_score >= 0.6.

Respond ONLY with valid JSON matching this schema:
{
  "generated_title": "string",
  "summary": "string",
  "clickbait_score": 0.0,
  "is_clickbait": false
}`;

interface SummaryResult {
  generated_title: string;
  summary: string;
  clickbait_score: number;
  is_clickbait: boolean;
}

export async function summarizeArticle(
  originalTitle: string,
  content: string,
  url: string
): Promise<SummaryResult> {
  const cacheKey = makeCacheKey('summary', { url });
  const cached = cacheGet<SummaryResult>(cacheKey);
  if (cached) return cached;

  if (!process.env.GEMINI_API_KEY) {
    console.warn('GEMINI_API_KEY not set — skipping AI summarization');
    const clean = content.replace(/<[^>]*>/g, ' ').replace(/\s+/g, ' ').trim();
    const result: SummaryResult = {
      generated_title: originalTitle,
      summary:
        clean.length > 200 ? clean.slice(0, 200) + '...' : clean || '',
      clickbait_score: 0,
      is_clickbait: false,
    };
    cacheSet(cacheKey, result, 3600000);
    return result;
  }

  const client = getGeminiClient();
  const model = process.env.GEMINI_MODEL || 'gemini-2.0-flash';

  try {
    const resp = await client.chat.completions.create({
      model,
      messages: [
        { role: 'system', content: SYSTEM_PROMPT },
        {
          role: 'user',
          content: `Original title: ${originalTitle}\n\nArticle content:\n${content.slice(0, 3000)}`,
        },
      ],
      temperature: 0.3,
      max_tokens: 500,
    });

    let raw = resp.choices[0]?.message?.content?.trim() || '';
    // Handle markdown code blocks
    if (raw.startsWith('```')) {
      raw = raw.split('\n').slice(1).join('\n').replace(/```$/, '').trim();
    }

    const result: SummaryResult = JSON.parse(raw);
    cacheSet(cacheKey, result, 3600000);
    return result;
  } catch (e) {
    console.error(`Summarization failed for "${url}":`, e);
    const clean = content.replace(/<[^>]*>/g, ' ').replace(/\s+/g, ' ').trim();
    return {
      generated_title: originalTitle,
      summary:
        clean.length > 200 ? clean.slice(0, 200) + '...' : clean || '',
      clickbait_score: 0,
      is_clickbait: false,
    };
  }
}

export async function batchSummarize(
  articles: Article[]
): Promise<Article[]> {
  const BATCH_SIZE = 5;
  const enriched: Article[] = [];

  for (let i = 0; i < articles.length; i += BATCH_SIZE) {
    const batch = articles.slice(i, i + BATCH_SIZE);
    const results = await Promise.all(
      batch.map(async (article) => {
        let content = article.content_snippet || '';
        if (!content || content.length < 50) {
          content = await extractArticleContent(article.original_url);
          article.content_snippet = content;
        }

        const result = await summarizeArticle(
          article.original_title,
          content,
          article.original_url
        );

        return {
          ...article,
          generated_title: result.generated_title,
          summary: result.summary,
          clickbait_score: result.clickbait_score,
          is_clickbait: result.is_clickbait,
        };
      })
    );
    enriched.push(...results);
  }

  return enriched;
}

File: pusheen/closed_loop/backend/services/summarizer.py
"""
AI-powered summarization & anti-clickbait title generation.

Uses OpenAI-compatible API (works with OpenAI, OpenRouter, local models).
All results are cached so repeat views are instant.
"""
import json
import logging

from openai import AsyncOpenAI

from ..core.cache import summary_cache
from ..core.config import settings

logger = logging.getLogger(__name__)

_openai_client: AsyncOpenAI | None = None


def _get_openai() -> AsyncOpenAI:
    global _openai_client
    if _openai_client is None:
        _openai_client = AsyncOpenAI(
            api_key=settings.OPENAI_API_KEY,
            base_url=settings.OPENAI_BASE_URL,
        )
    return _openai_client


SYSTEM_PROMPT = """You are a news analyst. Given an article's original title and content, produce:

1. **generated_title** — A concise, accurate, non-clickbait title that faithfully represents the article's actual content. Remove sensationalism, vague teasers, and emotional manipulation. Keep it informative.
2. **summary** — A TL;DR of 2-3 sentences capturing the key facts. No fluff.
3. **clickbait_score** — A float from 0.0 (not clickbait) to 1.0 (extreme clickbait) rating how clickbaity the ORIGINAL title is.
4. **is_clickbait** — Boolean, true if clickbait_score >= 0.6.

Respond ONLY with valid JSON matching this schema:
{
  "generated_title": "string",
  "summary": "string",
  "clickbait_score": 0.0,
  "is_clickbait": false
}"""


async def summarize_article(original_title: str, content: str, url: str) -> dict:
    """Generate summary, anti-clickbait title, and clickbait score."""
    cache_params = {"url": url}
    cached = summary_cache.get("summary", cache_params)
    if cached is not None:
        return cached

    if not settings.OPENAI_API_KEY:
        # Fallback when no API key configured
        result = {
            "generated_title": original_title,
            "summary": content[:200] + "..." if len(content) > 200 else content,
            "clickbait_score": 0.0,
            "is_clickbait": False,
        }
        summary_cache.set("summary", result, cache_params)
        return result

    client = _get_openai()
    user_msg = f"Original title: {original_title}\n\nArticle content:\n{content[:3000]}"

    try:
        resp = await client.chat.completions.create(
            model=settings.SUMMARIZATION_MODEL,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": user_msg},
            ],
            temperature=0.3,
            max_tokens=500,
        )
        raw = resp.choices[0].message.content.strip()
        # Parse JSON from response (handle markdown code blocks)
        if raw.startswith("```"):
            raw = raw.split("\n", 1)[1].rsplit("```", 1)[0]
        result = json.loads(raw)
        summary_cache.set("summary", result, cache_params)
        return result
    except Exception as e:
        logger.error(f"Summarization failed for '{url}': {e}")
        return {
            "generated_title": original_title,
            "summary": content[:200] + "..." if len(content) > 200 else content,
            "clickbait_score": 0.0,
            "is_clickbait": False,
        }


async def batch_summarize(articles: list[dict]) -> list[dict]:
    """Enrich a list of article dicts with summaries and generated titles."""
    import asyncio

    async def _process(article: dict) -> dict:
        content = article.get("content_snippet", "")
        if not content or len(content) < 50:
            # Try to fetch content from URL
            from .news_fetcher import extract_article_content
            content = await extract_article_content(article["original_url"])
            article["content_snippet"] = content

        result = await summarize_article(
            article["original_title"],
            content,
            article["original_url"],
        )
        article["generated_title"] = result.get("generated_title", article["original_title"])
        article["summary"] = result.get("summary", "")
        article["clickbait_score"] = result.get("clickbait_score", 0.0)
        article["is_clickbait"] = result.get("is_clickbait", False)
        return article

    # Process in batches of 5 to avoid rate limits
    batch_size = 5
    enriched = []
    for i in range(0, len(articles), batch_size):
        batch = articles[i : i + batch_size]
        results = await asyncio.gather(*[_process(a) for a in batch])
        enriched.extend(results)

    return enriched
